### PR TITLE
Docs Generation Infrastructure.

### PR DESCRIPTION
**Pull request by georgend**
Creates a Docs PR with each PR, and Service Module Generation.
Branch with Generated Docs: https://github.com/microsoftgraph/msgraph-sdk-powershell/tree/DocsGeneration/202107081545

### CODE REVIEW

File: config/ModulesMapping.jsonc [hunk 1]
```diff
@@ -1,40 +1,40 @@
 {
-    "Applications": "^applicationTemplates\\.|^applications\\.|^servicePrincipals\\.|^onPremisesPublishingProfiles\\.|^users.appRoleAssignment$|^groups.appRoleAssignment$",
+   //"Applications": "^applicationTemplates\\.|^applications\\.|^servicePrincipals\\.|^onPremisesPublishingProfiles\\.|^users.appRoleAssignment$|^groups.appRoleAssignment$",
```
Comment (peombwa) [inline config/ModulesMapping.jsonc:2]: Let's uncomment these. We want to generate all modules,

File: .azure-pipelines/common-templates/install-tools-template.yml [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+      displayName: 'Use .NET Core SDK 3.x'
+      inputs:
+        debugMode: false
+        version: 3.x
+
+    - task: UseDotNet@2
+      displayName: 'Use .NET Core SDK 5.x'
+      inputs:
+        debugMode: false
+        version: 5.x
```
Comment (peombwa) [inline .azure-pipelines/common-templates/install-tools-template.yml:21]: Do we need all these versions of .NET Core? Can we settle on just one version to help isolate .NET related issues in the future?

File: .azure-pipelines/generate-auth-module-template.yml [hunk 3]
```diff
@@ -35,7 +35,7 @@ jobs:
   displayName: Microsoft Graph PowerShell SDK Auth Module Generation
 
   steps:
-  - template: ./install-tools-template.yml
```
Comment (peombwa) [inline .azure-pipelines/generate-auth-module-template.yml:38]: Also delete `./install-tools-template.yml` here https://github.com/microsoftgraph/msgraph-sdk-powershell/blob/dev/.azure-pipelines/install-tools-template.yml.

File: .azure-pipelines/generation-templates/generate-helpdocs-template.yml [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+
+- task: PowerShell@2
+  displayName: Generate Help Docs
+  continueOnError: false
+  enabled: true
+  inputs:
+    targetType: 'inline'
+    pwsh: true
+    script: |  
+      $Modules = "${{ parameters.ModulesToGenerate }}" -split " "
```
Comment (peombwa) [inline .azure-pipelines/generation-templates/generate-helpdocs-template.yml:44]: `$Modules` variable is never used anywhere.

File: tools/GenerateHelp.ps1 [hunk 5]
```diff
@@ -0,0 +1,62 @@
+								 
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+Param(
+    $ModulesToGenerate = @(),
+    [string] $ModuleMappingConfigPath = (Join-Path $PSScriptRoot "..\config\ModulesMapping.jsonc")
+)
+Install-Module PlatyPS -Force
```
Comment (peombwa) [inline tools/GenerateHelp.ps1:8]: Add a check to see if PlatyPS is already installed before installing it.

File: .azure-pipelines/common-templates/install-tools-template.yml (hunk 2 above)
Comment (peombwa) [reply .azure-pipelines/common-templates/install-tools-template.yml:21]: It's also worth considering the End of Life (EOL) for some of these versions. See https://dotnet.microsoft.com/platform/support/policy/dotnet-core.

File: .azure-pipelines/common-templates/install-tools-template.yml [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    - task: NuGetToolInstaller@1
+      displayName: 'Install Nuget 5.7'
+      inputs:
+         versionSpec: 5.7.0
+         checkLatest: false # Optional
+
+    - task: NuGetAuthenticate@0
+
+    - task: PowerShell@2
+      displayName: 'Install Powershell Core'
```
Comment (peombwa) [inline .azure-pipelines/common-templates/install-tools-template.yml:32]: There is no need to install PowerShell core on an agent. To use PowerShell (core), use `pwsh: true` attribute on the `PowerShell@2` task or the `pwsh` shortcut task. e.g.,
``` yml
- task: PowerShell@2
    inputs:
      targetType: 'inline'
      script: 'Write-Host "I'm running on pwsh core!'
      pwsh: true

# Or

- pwsh: 'Write-Host "I'm running on pwsh core!"'
```

File: config/ModulesMapping.jsonc (hunk 1 above)
Comment (georgend) [reply config/ModulesMapping.jsonc:2]: Was for experimental purposes. Will be removed.

File: .azure-pipelines/common-templates/install-tools-template.yml (hunk 2 above)
Comment (georgend) [reply .azure-pipelines/common-templates/install-tools-template.yml:21]: Some of the security tools require those older versions of dotnet. 
Also its specified as `2.x` which picks the appropriate version.

Comment (georgend) [reply .azure-pipelines/common-templates/install-tools-template.yml]: https://docs.microsoft.com/en-us/azure/devops/pipelines/tasks/tool/dotnet-core-tool-installer?view=azure-devops#task-inputs
2.x => Install latest in major version.

File: .azure-pipelines/generation-templates/generate-service-modules.yml [hunk 7]
```diff
@@ -138,7 +142,12 @@ jobs:
           script: |
             [bool]$EnableSigning = if ("$(EnableSigning)" -eq "true") { $true } else { $false }
             $Modules = "$(ModulesToGenerate)" -split " "
-            . $(System.DefaultWorkingDirectory)/tools/GenerateModules.ps1 -Build -Test -UpdateAutoRest -EnableSigning:$EnableSigning -ModulesToGenerate $Modules
+            . $(System.DefaultWorkingDirectory)/tools/GenerateModules.ps1 -Build -Test -UpdateAutoRest -EnableSigning:$EnableSigning -ModulesToGenerate $Modules -ExcludeExampleTemplates
+       
+      - template: ./generation-templates/generate-helpdocs-template.yml
+        parameters:
+            DOCSBRANCH: $(DocsBranch)
+            ModulesToGenerate: $(ModulesToGenerate)
```
Comment (peombwa) [inline .azure-pipelines/generation-templates/generate-service-modules.yml:150]: Commit https://github.com/microsoftgraph/msgraph-sdk-powershell/pull/726/commits/43a48139dd7e48c3eb1c6132f19b73ac74832bb1 removed `ModulesToGenerate` parameter from `generate-helpdocs-template.yml`, yet this line still uses it. This change will break the weekly API refresh.

File: tools/GenerateHelp.ps1 [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+        Write-Host $v1ModuleDocs
+        Select-MgProfile v1.0
+        Update-MarkdownHelpModule -Path $v1ModuleDocs -RefreshModulePage -AlphabeticParamsOrder -ExcludeDontShow
+        git add $v1ModuleDocs
+    }
+    else {
+        Write-Warning "v1.0 Docs for $ModuleName not Found"
+    }
+    git status
+    git commit -m "Docs Generation for $ModuleName [skip ci]"
```
Comment (peombwa) [inline tools/GenerateHelp.ps1:62]: We should move all git commands in this file to one central file for maintainability, ideally in `.azure-pipelines/generation-templates/generate-helpdocs-template.yml`. This is the same approach we take with the weekly API refresh https://github.com/microsoftgraph/msgraph-sdk-powershell/blob/dev/.azure-pipelines/common-templates/download-openapi-docs.yml#L122.

Comment (georgend) [reply tools/GenerateHelp.ps1:62]: At this point we know the module (including its name), module directory and generated/updated docs. I create a commit per module where applicable. This information is not available with ease elsewhere.

File: .azure-pipelines/generation-templates/generate-service-modules.yml (hunk 7 above)
Comment (georgend) [reply .azure-pipelines/generation-templates/generate-service-modules.yml:150]: Good catch.

File: .azure-pipelines/generation-templates/generate-helpdocs-template.yml [hunk 9]
```diff
@@ -0,0 +1,57 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+parameters:
+  - name: DOCSBRANCH
+    type: string
+    default: "DocsGeneration"
+  - name: ModulesToGenerate
```
Comment (peombwa) [inline .azure-pipelines/generation-templates/generate-helpdocs-template.yml:8]: Parameter `ModulesToGenerate` is declared but never used anywhere in this template. Shouldn't we be passing it as a parameter to `. $(System.DefaultWorkingDirectory)/tools/GenerateHelp.ps1`?

File: .azure-pipelines/generation-templates/generate-helpdocs-template.yml [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+
+- task: PowerShell@2
+  displayName: Generate Help Docs
+  continueOnError: false
+  enabled: true
+  inputs:
+    targetType: 'inline'
+    pwsh: true
+    script: |  
+      . $(System.DefaultWorkingDirectory)/tools/GenerateHelp.ps1
```
Comment (peombwa) [inline .azure-pipelines/generation-templates/generate-helpdocs-template.yml:44]: We should pass `$(ModulesToGenerate)` to `GenerateHelp.ps1`.
```suggestion
       $Modules = "$(ModulesToGenerate)" -split " "
      . $(System.DefaultWorkingDirectory)/tools/GenerateHelp.ps1 -ModulesToGenerate $Modules
```

File: .azure-pipelines/generation-templates/generate-helpdocs-template.yml (hunk 9 above)
Comment (peombwa) [reply .azure-pipelines/generation-templates/generate-helpdocs-template.yml:8]: See suggestion -> https://github.com/microsoftgraph/msgraph-sdk-powershell/pull/726#discussion_r662648181.

**Comment by peombwa** [comment]
@georgend it looks like we've got an error generating help docs for `Get-MgGroupCreatedOnBehalfOf.md`, see AzDO logs:
``` powershell
ForEach-Object : Exception calling "NodeModelToMamlModel" with "1" argument(s): "C:\a\1\s\src\Groups\Groups\docs\v1.0-beta\Get-MgGroupCreatedOnBehalfOf.md:104:(342) '
### Get-MgGroupCreatedOnBehalf
'
 Expect Heading level 2"
At C:\Users\AzDevOps\Documents\PowerShell\Modules\PlatyPS\0.14.2\platyPS.psm1:1598 char:22
+     $markdownFiles | ForEach-Object {
+                      ~~~~~~~~~~~~~~~~
+ CategoryInfo          : NotSpecified: (:) [ForEach-Object], MethodInvocationException
+ FullyQualifiedErrorId : HelpSchemaException,Microsoft.PowerShell.Commands.ForEachObjectCommand

```

**Comment by georgend** [comment]
> @georgend it looks like we've got an error generating help docs for `Get-MgGroupCreatedOnBehalfOf.md`, see AzDO logs:
> 
> ```powershell
> ForEach-Object : Exception calling "NodeModelToMamlModel" with "1" argument(s): "C:\a\1\s\src\Groups\Groups\docs\v1.0-beta\Get-MgGroupCreatedOnBehalfOf.md:104:(342) '
> ### Get-MgGroupCreatedOnBehalf
> '
>  Expect Heading level 2"
> At C:\Users\AzDevOps\Documents\PowerShell\Modules\PlatyPS\0.14.2\platyPS.psm1:1598 char:22
> +     $markdownFiles | ForEach-Object {
> +                      ~~~~~~~~~~~~~~~~
> + CategoryInfo          : NotSpecified: (:) [ForEach-Object], MethodInvocationException
> + FullyQualifiedErrorId : HelpSchemaException,Microsoft.PowerShell.Commands.ForEachObjectCommand
> ```

Issue resolved.

**Review by peombwa — APPROVED** [review]

**Pull request merged**